Support NVFP4 quantized dense models on AMD CDNA2/CDNA3 GPUs

haohui opened the pull request:
<!-- Thank you for your contribution! We appreciate it. The following guidelines will help improve your pull request and facilitate feedback. If anything is unclear, don't hesitate to submit your pull request and ask the maintainers for assistance. -->

## Motivation

This PR leverages [Petit](https://github.com/causalflow-ai/petit-kernel) to efficiently emulate FP4 supports on AMD CDNA2 / CDNA3 GPUs, so that they can run FP4 quantized models such as [Llama-3.3-70B-Instruct-FP4](https://huggingface.co/nvidia/Llama-3.3-70B-Instruct-FP4)

<!-- Explain the purpose of this PR and the goals it aims to achieve. -->

## Modifications

<!-- Describe the changes made in this PR. -->

This PR adds a new quantization scheme, `petit_fp4` to support fp4 models quantized by ModelOpt.

## Benchmarks

The MMLU benchmark scores 82.22 and 78.88 on the [Llama-3.3-70B-Instruct](https://github.com/causalflow-ai/petit-kernel/blob/master/meta-llama/Llama-3.3-70B-Instruct) / [Llama-3.3-70B-Instruct-FP4](https://huggingface.co/nvidia/Llama-3.3-70B-Instruct-FP4) models respectively using a server with 8x AMD MI300 GPUs. The scores are aligned with the scores provided [here](https://huggingface.co/nvidia/Llama-3.3-70B-Instruct-FP4).

We also evaluated the offline generation performance of Llama-3-8B on a MI250x GPU for both the original BF16 and quantized FP4 version:

Serving Llama-3-8B BF16 on a MI250x GPU:
```
====== Offline Throughput Benchmark Result =======
Backend:                                 engine
Successful requests:                     10
Benchmark duration (s):                  12.88
Total input tokens:                      1960
Total generated tokens:                  2774
Last generation throughput (tok/s):      44.33
Request throughput (req/s):              0.78
Input token throughput (tok/s):          152.17
Output token throughput (tok/s):         215.36
Total token throughput (tok/s):          367.53
==================================================
```

Serving Llama-3-8B, FP4 quantized on a MI250x GPU:
```
====== Offline Throughput Benchmark Result =======
Backend:                                 engine
Successful requests:                     10
Benchmark duration (s):                  10.33
Total input tokens:                      1960
Total generated tokens:                  2774
Last generation throughput (tok/s):      64.73
Request throughput (req/s):              0.97
Input token throughput (tok/s):          189.78
Output token throughput (tok/s):         268.59
Total token throughput (tok/s):          458.37
==================================================
```

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [x] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [x] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [x] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [x] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [x] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

haohui commented:
I have tested the PR with MMLU on the AMD MI300x platform.

The [Llama-3.3-70B-Instruct](https://github.com/causalflow-ai/petit-kernel/blob/master/meta-llama/Llama-3.3-70B-Instruct) / [Llama-3.3-70B-Instruct-FP4](https://huggingface.co/nvidia/Llama-3.3-70B-Instruct-FP4) model gets MMLU scores of 82.22 and 78.88 respectively.

HaiShaw reviewed (changes requested):
Can we change `petit_fp4` to `petit_nvfp4`?

haohui commented:
Update the PR to address comments.

@HaiShaw The drops are expected due to the loss of accuracy in quantizations.

HaiShaw reviewed (changes requested):
Few more name changes?

haohui commented:
I suspect it might be subtle differences how the evaluation is set up. Trying to find a B200 machine to test it out...

haohui committed:
Support FP4 quantized models on AMD CDNA2/CDNA3 GPUs

haohui commented:
Updated the PR to address the comments. Also pull the petit-kernel 0.0.2 release. The 0.0.2 releases improves both the performance and accuracy.

I use `sglang.bench_offline_throughput` to measure the throughputs of serving the FP4-quantized and the original BF16 version of the Llama-3.3-70B-Instruct model.

```
MI300x 8x, FP4, prompts=64
====== Offline Throughput Benchmark Result =======
Backend:                                 engine
Successful requests:                     64
Benchmark duration (s):                  24.14
Total input tokens:                      19441
Total generated tokens:                  13499
Last generation throughput (tok/s):      103.80
Request throughput (req/s):              2.65
Input token throughput (tok/s):          805.28
Output token throughput (tok/s):         559.15
Total token throughput (tok/s):          1364.44
==================================================

MI300x 8x, BF16, prompts=64
====== Offline Throughput Benchmark Result =======
Backend:                                 engine
Successful requests:                     64
Benchmark duration (s):                  22.73
Total input tokens:                      19441
Total generated tokens:                  13499
Last generation throughput (tok/s):      97.33
Request throughput (req/s):              2.82
Input token throughput (tok/s):          855.24
Output token throughput (tok/s):         593.84
Total token throughput (tok/s):          1449.08
==================================================

MI300x 8x, FP4, prompts=10
====== Offline Throughput Benchmark Result =======
Backend:                                 engine
Successful requests:                     10
Benchmark duration (s):                  8.38
Total input tokens:                      1960
Total generated tokens:                  2774
Last generation throughput (tok/s):      111.09
Request throughput (req/s):              1.19
Input token throughput (tok/s):          234.02
Output token throughput (tok/s):         331.20
Total token throughput (tok/s):          565.22
==================================================

MI300x 8x, BF16, prompts=10
====== Offline Throughput Benchmark Result =======
Backend:                                 engine
Successful requests:                     10
Benchmark duration (s):                  8.67
Total input tokens:                      1960
Total generated tokens:                  2774
Last generation throughput (tok/s):      99.96
Request throughput (req/s):              1.15
Input token throughput (tok/s):          226.10
Output token throughput (tok/s):         320.00
Total token throughput (tok/s):          546.10
==================================================
```

Thanks for @HaiShaw pointing out the discrepancy of the MMLU scores -- I've fixed the bugs in the new release. I ran lm_eval and sglang-v0.4.9.post2 on a server with 8x MI300x and a server with 8x NVIDIA B200. Here are the MMLU scores with batch size equals 128:

| | BF16 | FP4 |
|--- | ----- | ----- |
| MI300x | 82.15 | 80.79 |
| B200 | 82.17 | 80.29 |

haohui commented:
For full reproducibility, here are the commands to try out the PR:

```
import sglang as sgl
engine = sgl.Engine(model_path='nvidia/Llama-3.3-70B-Instruct-FP4')
engine.generate('hello')
```

Command lines to compute the MMLU score on 8x MI300X:

```
python3 -m lm_eval --tasks mmlu --model sglang --model_args pretrained=nvidia/Llama-3.3-70B-Instruct,context_length=3072,tp_size=8,attention_backend=triton,mem_fraction_static=0.5 --num_fewshot 5 --batch_size 128 
```

Command lines to evaluate the offline performance:

```
python3 -m sglang.bench_offline_throughput --model-path nvidia/Llama-3.3-70B-Instruct-FP4 --tensor-parallel-size 8 --num-prompts 10
python3 -m sglang.bench_offline_throughput --model-path nvidia/Llama-3.3-70B-Instruct-FP4 --tensor-parallel-size 8 --num-prompts 64
```

HaiShaw reviewed (approved):
LGTM, thanks!

haohui committed:
Fix rebase issues

HaiShaw commented:
@haohui petit_kernel package works to both AMD and NV?
@saienduri there is a change to pyproject, check CI please.

haohui committed:
Merge branch 'main' into amd-mi300-fp4

haohui commented:
Seems like the failed tests are flaky due to timeouts / network issues. @HaiShaw can you please shed some lights on how to proceed?

HaiShaw committed:
Merge branch 'main' into amd-mi300-fp4

saienduri committed:
Merge branch 'main' into amd-mi300-fp4

haohui committed:
Only install petit_kernel for ROCm

haohui committed:
Merge branch 'main' into amd-mi300-fp4

haohui committed:
Guard against non-ROCm set up

haohui commented:
We have some plans but an optimized MoE would take quite a bit of work. Are there any use cases now?

RioXu commented:
@haohui We are soon getting a rig of 4x MI250, which is hopefully capable of running some latest models (quantized) like GLM-5 or qwen 3.5 397B, but I've never seen any use case with the same setup on the Internet. Would you consider adding support for nvfp4 MoE?
